File: engine_financeiro.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Engine Financeiro - Core financial processing system
Processes CRM sales and updates financial schema with automatic cost allocation
"""

import json
import os
from datetime import datetime
from typing import Dict, Any


class EngineFinanceiro:
    """Financial engine for processing sales transactions"""
    
    SCHEMA_PATH = 'schema_financeiro.json'
    LOG_PATH = 'transacoes.log'
    
    # Business constants
    RECEITA_FIXA_POR_VENDA = 597.00
    TAXA_IMPOSTO = 0.16  # 16%
    
    # Unit COGS per sale
    COGS = {
        'Servidor': 10.00,
        'Tokens GPT': 10.00,
        'Telefone': 30.00,
        'API Oficial Whatsapp': 200.00
    }
    
    def __init__(self):
        """Initialize the financial engine"""
        self.schema = self.load_schema()
    
    def load_schema(self) -> Dict[str, Any]:
        """Load financial schema from JSON file"""
        if not os.path.exists(self.SCHEMA_PATH):
            raise FileNotFoundError(f"Schema file not found: {self.SCHEMA_PATH}")
        
        with open(self.SCHEMA_PATH, 'r', encoding='utf-8') as f:
            return json.load(f)
    
    def save_schema(self):
        """Save updated schema back to JSON file"""
        with open(self.SCHEMA_PATH, 'w', encoding='utf-8') as f:
            json.dump(self.schema, f, indent=2, ensure_ascii=False)
    
    def find_category(self, category_list: list, label: str) -> Dict[str, Any]:
        """Find a category by label in a list"""
        for item in category_list:
            if label.lower() in item['label'].lower():
                return item
        return None
    
    def registrar_nova_venda(self, valor_comissao_recuperada: float):
        """
        Process a new sale transaction
        
        Args:
            valor_comissao_recuperada: Variable commission amount from recovered sales
        """
        # Calculate totals
        receita_total = self.RECEITA_FIXA_POR_VENDA + valor_comissao_recuperada
        total_cogs = sum(self.COGS.values())
        impostos = receita_total * self.TAXA_IMPOSTO
        saldo_liquido = receita_total - total_cogs - impostos
        
        # Update Fixed Revenue
        receita_fixa = self.find_category(
            self.schema['receitas']['fixa'], 
            'Receita Fixa - Mensalidade'
        )
        if receita_fixa:
            receita_fixa['REAL_2026'] += self.RECEITA_FIXA_POR_VENDA
        
        # Update Variable Revenue
        receita_variavel = self.find_category(
            self.schema['receitas']['variavel'],
            'Receita Variável - Comissão'
        )
        if receita_variavel:
            receita_variavel['REAL_2026'] += valor_comissao_recuperada
        
        # Update Taxes
        impostos_item = self.find_category(
            self.schema['impostos'],
            'Impostos'
        )
        if impostos_item:
            impostos_item['REAL_2026'] += impostos
        
        # Update COGS (Variable Expenses)
        for cogs_name, cogs_value in self.COGS.items():
            cogs_item = self.find_category(
                self.schema['despesas_variaveis'],
                cogs_name
            )
            if cogs_item:
                cogs_item['REAL_2026'] += cogs_value
        
        # Save updated schema
        self.save_schema()
        
        # Log transaction
        self.log_transacao(
            receita_total=receita_total,
            total_cogs=total_cogs,
            impostos=impostos,
            saldo_liquido=saldo_liquido
        )
        
        # Print summary
        print(f"\n{'='*70}")
        print(f"✅ VENDA REGISTRADA")
        print(f"{'='*70}")
        print(f"Receita Fixa:              R$ {self.RECEITA_FIXA_POR_VENDA:>10.2f}")
        print(f"Receita Variável (Comissão): R$ {valor_comissao_recuperada:>10.2f}")
        print(f"                           {'─'*30}")
        print(f"Receita Total:             R$ {receita_total:>10.2f}")
        print(f"\nCustos Variáveis (COGS):")
        for cogs_name, cogs_value in self.COGS.items():
            print(f"  - {cogs_name:<25} R$ {cogs_value:>10.2f}")
        print(f"                           {'─'*30}")
        print(f"Total COGS:                R$ {total_cogs:>10.2f}")
        print(f"Impostos (16%):            R$ {impostos:>10.2f}")
        print(f"                           {'─'*30}")
        print(f"Resultado Bruto:           R$ {receita_total - total_cogs:>10.2f}")
        print(f"Saldo Líquido:             R$ {saldo_liquido:>10.2f}")
        print(f"{'='*70}\n")
        
        return {
            'receita_total': receita_total,
            'total_cogs': total_cogs,
            'impostos': impostos,
            'saldo_liquido': saldo_liquido
        }
    
    def log_transacao(self, receita_total: float, total_cogs: float, 
                     impostos: float, saldo_liquido: float):
        """
        Log transaction to audit file
        
        Args:
            receita_total: Total revenue for this sale
            total_cogs: Total cost of goods sold
            impostos: Tax amount
            saldo_liquido: Net profit after costs and taxes
        """
        timestamp = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
        log_line = (
            f"{timestamp} | "
            f"Venda: R$ {receita_total:>8.2f} | "
            f"COGS: R$ {total_cogs:>6.2f} | "
            f"Impostos: R$ {impostos:>8.2f} | "
            f"Líquido: R$ {saldo_liquido:>8.2f}\n"
        )
        
        with open(self.LOG_PATH, 'a', encoding='utf-8') as f:
            f.write(log_line)
    
    def get_total_metrics(self) -> Dict[str, float]:
        """Get current total metrics from schema"""
        total_receita_fixa = sum(
            item['REAL_2026'] for item in self.schema['receitas']['fixa']
        )
        total_receita_variavel = sum(
            item['REAL_2026'] for item in self.schema['receitas']['variavel']
        )
        total_impostos = sum(
            item['REAL_2026'] for item in self.schema['impostos']
        )
        total_cogs = sum(
            item['REAL_2026'] for item in self.schema['despesas_variaveis']
        )
        
        return {
            'receita_fixa': total_receita_fixa,
            'receita_variavel': total_receita_variavel,
            'receita_total': total_receita_fixa + total_receita_variavel,
            'impostos': total_impostos,
            'cogs': total_cogs,
            'resultado_bruto': total_receita_fixa + total_receita_variavel - total_cogs,
            'lucro_liquido': total_receita_fixa + total_receita_variavel - total_cogs - total_impostos
        }


# ============================================================================
# TEST SIMULATION - 5 New Customers
# ============================================================================

if __name__ == "__main__":
    print("\n" + "="*70)
    print("🚀 SIMULAÇÃO DE VENDAS - ENGINE FINANCEIRO")
    print("="*70)
    
    # Initialize engine
    engine = EngineFinanceiro()
    
    # Show initial state
    print("\n📊 ESTADO INICIAL:")
    metrics = engine.get_total_metrics()
    print(f"Receita Total Atual: R$ {metrics['receita_total']:.2f}")
    print(f"COGS Total Atual: R$ {metrics['cogs']:.2f}")
    print(f"Lucro Líquido Atual: R$ {metrics['lucro_liquido']:.2f}")
    
    # Simulate 5 new customers with varying commission values
    clientes = [
        {'nome': 'Cliente 1', 'comissao': 150.00},
        {'nome': 'Cliente 2', 'comissao': 200.00},
        {'nome': 'Cliente 3', 'comissao': 180.00},
        {'nome': 'Cliente 4', 'comissao': 220.00},
        {'nome': 'Cliente 5', 'comissao': 175.00},
    ]
    
    print(f"\n{'='*70}")
    print(f"🎯 PROCESSANDO {len(clientes)} NOVOS CLIENTES")
    print(f"{'='*70}")
    
    for idx, cliente in enumerate(clientes, 1):
        print(f"\n[{idx}/{len(clientes)}] Processando {cliente['nome']}...")
        engine.registrar_nova_venda(cliente['comissao'])
    
    # Show final state
    print(f"\n{'='*70}")
    print("📈 ESTADO FINAL:")
    print(f"{'='*70}")
    
    final_metrics = engine.get_total_metrics()
    print(f"\nReceita Fixa Total:        R$ {final_metrics['receita_fixa']:>10.2f}")
    print(f"Receita Variável Total:    R$ {final_metrics['receita_variavel']:>10.2f}")
    print(f"                           {'─'*30}")
    print(f"Receita Total:             R$ {final_metrics['receita_total']:>10.2f}")
    print(f"\nCOGS Total:                R$ {final_metrics['cogs']:>10.2f}")
    print(f"Impostos Total:            R$ {final_metrics['impostos']:>10.2f}")
    print(f"                           {'─'*30}")
    print(f"Resultado Bruto:           R$ {final_metrics['resultado_bruto']:>10.2f}")
    print(f"Lucro Líquido:             R$ {final_metrics['lucro_liquido']:>10.2f}")
    print(f"\n{'='*70}")
    print(f"✅ Simulação concluída! Confira transacoes.log para auditoria.")
    print(f"{'='*70}\n")

File: engine_financeiro.js
#!/usr/bin/env node
/**
 * Engine Financeiro - Core financial processing system
 * Processes CRM sales and updates financial schema with automatic cost allocation
 */

import fs from 'fs';
import path from 'path';

class EngineFinanceiro {
    // Business constants
    static RECEITA_FIXA_POR_VENDA = 597.00;
    static TAXA_IMPOSTO = 0.16; // 16%

    // Unit COGS per sale
    static COGS = {
        'Servidor': 10.00,
        'Tokens GPT': 10.00,
        'Telefone': 30.00,
        'API Oficial Whatsapp': 200.00
    };

    constructor(schemaPath = 'schema_financeiro.json', logPath = 'transacoes.log') {
        this.schemaPath = schemaPath;
        this.logPath = logPath;
        this.schema = this.loadSchema();
    }

    /**
     * Load financial schema from JSON file
     */
    loadSchema() {
        if (!fs.existsSync(this.schemaPath)) {
            throw new Error(`Schema file not found: ${this.schemaPath}`);
        }

        const data = fs.readFileSync(this.schemaPath, 'utf-8');
        return JSON.parse(data);
    }

    /**
     * Save updated schema back to JSON file
     */
    saveSchema() {
        fs.writeFileSync(
            this.schemaPath,
            JSON.stringify(this.schema, null, 2),
            'utf-8'
        );
    }

    /**
     * Find a category by label in a list
     */
    findCategory(categoryList, label) {
        return categoryList.find(item =>
            item.label.toLowerCase().includes(label.toLowerCase())
        );
    }

    /**
     * Process a new sale transaction
     * @param {number} valorComissaoRecuperada - Variable commission amount
     */
    registrarNovaVenda(valorComissaoRecuperada) {
        // Calculate totals
        const receitaTotal = EngineFinanceiro.RECEITA_FIXA_POR_VENDA + valorComissaoRecuperada;
        const totalCogs = Object.values(EngineFinanceiro.COGS).reduce((a, b) => a + b, 0);
        const impostos = receitaTotal * EngineFinanceiro.TAXA_IMPOSTO;
        const saldoLiquido = receitaTotal - totalCogs - impostos;

        // Update Fixed Revenue
        const receitaFixa = this.findCategory(
            this.schema.receitas.fixa,
            'Receita Fixa - Mensalidade'
        );
        if (receitaFixa) {
            receitaFixa.REAL_2026 += EngineFinanceiro.RECEITA_FIXA_POR_VENDA;
        }

        // Update Variable Revenue
        const receitaVariavel = this.findCategory(
            this.schema.receitas.variavel,
            'Receita Variável - Comissão'
        );
        if (receitaVariavel) {
            receitaVariavel.REAL_2026 += valorComissaoRecuperada;
        }

        // Update Taxes
        const impostosItem = this.findCategory(
            this.schema.impostos,
            'Impostos'
        );
        if (impostosItem) {
            impostosItem.REAL_2026 += impostos;
        }

        // Update COGS (Variable Expenses)
        for (const [cogsName, cogsValue] of Object.entries(EngineFinanceiro.COGS)) {
            const cogsItem = this.findCategory(
                this.schema.despesas_variaveis,
                cogsName
            );
            if (cogsItem) {
                cogsItem.REAL_2026 += cogsValue;
            }
        }

        // Save updated schema
        this.saveSchema();

        // Log transaction
        this.logTransacao(receitaTotal, totalCogs, impostos, saldoLiquido);

        // Print summary
        console.log(`\n${'='.repeat(70)}`);
        console.log(`✅ VENDA REGISTRADA`);
        console.log(`${'='.repeat(70)}`);
        console.log(`Receita Fixa:              R$ ${EngineFinanceiro.RECEITA_FIXA_POR_VENDA.toFixed(2).padStart(10)}`);
        console.log(`Receita Variável (Comissão): R$ ${valorComissaoRecuperada.toFixed(2).padStart(10)}`);
        console.log(`                           ${'─'.repeat(30)}`);
        console.log(`Receita Total:             R$ ${receitaTotal.toFixed(2).padStart(10)}`);
        console.log(`\nCustos Variáveis (COGS):`);
        for (const [cogsName, cogsValue] of Object.entries(EngineFinanceiro.COGS)) {
            console.log(`  - ${cogsName.padEnd(25)} R$ ${cogsValue.toFixed(2).padStart(10)}`);
        }
        console.log(`                           ${'─'.repeat(30)}`);
        console.log(`Total COGS:                R$ ${totalCogs.toFixed(2).padStart(10)}`);
        console.log(`Impostos (16%):            R$ ${impostos.toFixed(2).padStart(10)}`);
        console.log(`                           ${'─'.repeat(30)}`);
        console.log(`Resultado Bruto:           R$ ${(receitaTotal - totalCogs).toFixed(2).padStart(10)}`);
        console.log(`Saldo Líquido:             R$ ${saldoLiquido.toFixed(2).padStart(10)}`);
        console.log(`${'='.repeat(70)}\n`);

        return {
            receitaTotal,
            totalCogs,
            impostos,
            saldoLiquido
        };
    }

    /**
     * Log transaction to audit file
     */
    logTransacao(receitaTotal, totalCogs, impostos, saldoLiquido) {
        const timestamp = new Date().toLocaleString('pt-BR', {
            year: 'numeric',
            month: '2-digit',
            day: '2-digit',
            hour: '2-digit',
            minute: '2-digit',
            second: '2-digit',
            hour12: false
        });

        const logLine =
            `${timestamp} | ` +
            `Venda: R$ ${receitaTotal.toFixed(2).padStart(8)} | ` +
            `COGS: R$ ${totalCogs.toFixed(2).padStart(6)} | ` +
            `Impostos: R$ ${impostos.toFixed(2).padStart(8)} | ` +
            `Líquido: R$ ${saldoLiquido.toFixed(2).padStart(8)}\n`;

        fs.appendFileSync(this.logPath, logLine, 'utf-8');
    }

    /**
     * Get current total metrics from schema
     */
    getTotalMetrics() {
        const totalReceitaFixa = this.schema.receitas.fixa.reduce(
            (sum, item) => sum + item.REAL_2026, 0
        );
        const totalReceitaVariavel = this.schema.receitas.variavel.reduce(
            (sum, item) => sum + item.REAL_2026, 0
        );
        const totalImpostos = this.schema.impostos.reduce(
            (sum, item) => sum + item.REAL_2026, 0
        );
        const totalCogs = this.schema.despesas_variaveis.reduce(
            (sum, item) => sum + item.REAL_2026, 0
        );

        const receitaTotal = totalReceitaFixa + totalReceitaVariavel;
        const resultadoBruto = receitaTotal - totalCogs;
        const lucroLiquido = receitaTotal - totalCogs - totalImpostos;

        return {
            receitaFixa: totalReceitaFixa,
            receitaVariavel: totalReceitaVariavel,
            receitaTotal,
            impostos: totalImpostos,
            cogs: totalCogs,
            resultadoBruto,
            lucroLiquido
        };
    }
}

// ============================================================================
// TEST SIMULATION - 5 New Customers
// ============================================================================

console.log(`\n${'='.repeat(70)}`);
console.log(`🚀 SIMULAÇÃO DE VENDAS - ENGINE FINANCEIRO`);
console.log(`${'='.repeat(70)}`);

// Initialize engine
const engine = new EngineFinanceiro();

// Show initial state
console.log(`\n📊 ESTADO INICIAL:`);
let metrics = engine.getTotalMetrics();
console.log(`Receita Total Atual: R$ ${metrics.receitaTotal.toFixed(2)}`);
console.log(`COGS Total Atual: R$ ${metrics.cogs.toFixed(2)}`);
console.log(`Lucro Líquido Atual: R$ ${metrics.lucroLiquido.toFixed(2)}`);

// Simulate 5 new customers with varying commission values
const clientes = [
    { nome: 'Cliente 1', comissao: 150.00 },
    { nome: 'Cliente 2', comissao: 200.00 },
    { nome: 'Cliente 3', comissao: 180.00 },
    { nome: 'Cliente 4', comissao: 220.00 },
    { nome: 'Cliente 5', comissao: 175.00 },
];

console.log(`\n${'='.repeat(70)}`);
console.log(`🎯 PROCESSANDO ${clientes.length} NOVOS CLIENTES`);
console.log(`${'='.repeat(70)}`);

clientes.forEach((cliente, idx) => {
    console.log(`\n[${idx + 1}/${clientes.length}] Processando ${cliente.nome}...`);
    engine.registrarNovaVenda(cliente.comissao);
});

// Show final state
console.log(`\n${'='.repeat(70)}`);
console.log(`📈 ESTADO FINAL:`);
console.log(`${'='.repeat(70)}`);

const finalMetrics = engine.getTotalMetrics();
console.log(`\nReceita Fixa Total:        R$ ${finalMetrics.receitaFixa.toFixed(2).padStart(10)}`);
console.log(`Receita Variável Total:    R$ ${finalMetrics.receitaVariavel.toFixed(2).padStart(10)}`);
console.log(`                           ${'─'.repeat(30)}`);
console.log(`Receita Total:             R$ ${finalMetrics.receitaTotal.toFixed(2).padStart(10)}`);
console.log(`\nCOGS Total:                R$ ${finalMetrics.cogs.toFixed(2).padStart(10)}`);
console.log(`Impostos Total:            R$ ${finalMetrics.impostos.toFixed(2).padStart(10)}`);
console.log(`                           ${'─'.repeat(30)}`);
console.log(`Resultado Bruto:           R$ ${finalMetrics.resultadoBruto.toFixed(2).padStart(10)}`);
console.log(`Lucro Líquido:             R$ ${finalMetrics.lucroLiquido.toFixed(2).padStart(10)}`);
console.log(`\n${'='.repeat(70)}`);
console.log(`✅ Simulação concluída! Confira transacoes.log para auditoria.`);
console.log(`${'='.repeat(70)}\n`);
